flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,353 @@
+                }
+            } catch (OutOfMemoryError outOfMemoryError) {
+                int allocated = buffers.size();
+                buffers.clear();
+                throw new OutOfMemoryError(
+                        String.format(
+                                "Can't allocate enough direct buffer for batch shuffle read buffer "
+                                        + "pool (bytes allocated: %d, bytes still needed: %d). To "
+                                        + "avoid the exception, you need to do one of the following"
+                                        + " adjustments: 1) If you have ever decreased %s, you need"
+                                        + " to undo the decrement; 2) If you ever increased %s, you"
+                                        + " should also increase %s; 3) If neither the above cases,"
+                                        + " it usually means some other parts of your application "
+                                        + "have consumed too many direct memory and the value of %s"
+                                        + " should be increased.",
+                                allocated * bufferSize,
+                                (numTotalBuffers - allocated) * bufferSize,
+                                TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY.key(),
+                                TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key(),
+                                TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY.key(),
+                                TaskManagerOptions.TASK_OFF_HEAP_MEMORY.key()));
+            }
+        }
+    }
+
+    /**
+     * Requests a collection of buffers (determined by {@link #numBuffersPerRequest}) from this
+     * buffer pool. Exception will be thrown if no enough buffers can be allocated in the given
+     * timeout and the corresponding requester is not holding any allocated buffers currently.
+     */
+    public List<MemorySegment> requestBuffers(Object owner) throws Exception {
+        checkArgument(owner != null, "Owner must be not null.");
+
+        List<MemorySegment> allocated = new ArrayList<>(numBuffersPerRequest);
+        synchronized (buffers) {
+            checkState(!destroyed, "Buffer pool is already destroyed.");
+
+            if (!initialized) {
+                initialize();
+            }
+
+            long startTime = System.currentTimeMillis();
+            Counter counter = numBuffersAllocated.get(owner);
+            while (buffers.size() < numBuffersPerRequest
+                    || (counter != null
+                            && counter.get() + numBuffersPerRequest > maxBuffersPerRequester)) {
+                checkState(!destroyed, "Buffer pool is already destroyed.");
+
+                buffers.wait(requestTimeout);
+                counter = numBuffersAllocated.get(owner);
+
+                if (counter == null && System.currentTimeMillis() - startTime >= requestTimeout) {
```

Comment:
I have another concern here in that based on the condition for waiting, it seems a requester (owner) might have multiple threads work for it concurrently, right ? if so, suppose the current thread keep waiting due to another thread work for it holds too much buffer than `maxBuffersPerRequester` for a long time, and at last round, the other thread recycled all the buffers after timeout, then TimeoutException is thrown here. Would this also be expected ? It seems each requester has hold some buffers in this case and no requester are starved.

Reply:
It should be guaranteed only one thread can serve a requester at the same time, I can add this constraint to the java doc of this method.

Reply:
buffer recycle will always notify the waiting thread